# **LZ_Wrapper: A wrapper utility to create data, a local web server and html files for LocusZoom js to work.**

## **Requirements**

- Python
- Python Flask package
- Pandas
- Internet connection


## **Input data**

Three types of input files are needed for this utility to work:
 
### 1. For each trait/phenotype of interest, a tab-separated \<trait\>.txt file with all the loci of interest in it. 

For example, one would have a si.txt for smoking initiation, sc.txt for smoking cessation, and so on. These files **should have a header line**. The header line should start with a **#**.

  The important columns here are:
  - Column 1 : **Chromosome**
  - Column 2 : **Position**
  - Column 3 : **Marker/Variant ID**. The variant ID should be of the format **_chr:position_REF/ALT_**. The ref allele is inferred from this variant id.
  - Column 4 : **p value**
  - Column 5 : **Minor allele frequency (MAF)**
  The trait file can have other columns after column 5 - those will be ignored.
  See Examples for trait file examples (t1.txt, t2.txt, t3.txt).
  
### 2. For each phenotype, a tab-separated sig_variant.txt file with all the significant variants a user may want to explore (the TOP hits in the app for a phenotype). 

**This file should not have a header.**

For the example, this file contains the positions in TOP Hits in the following image:
 ![Alt text](/Example/TopHits.png?raw=true "Top Hits")

  There are two columns in the significant variants file:
  - Column 1 : **Chromosome**
  - Column 2 : **Position**
  An optional column can be added to the significant variants file. 
  - Optional Column 3 : **Variant name**. For variants with a rsid or a specific name, this column can be populated by the name. For variants without a specific name, do not add this column. 
  
### 3. A tab-separated phenotypes file. 

This file specifies the different phenotypes/traits of interest and their ids, and the specific significant_variant file.

  There are three columns in the phenotypes file:
  - Column 1 : **Phenotype id**
  - Column 2 : **Phenotype name**
  - Column 3 : **Significant_variant file name**
  
  **NOTE:** Be very careful to have the phenotype id exactly the same as the filenames for each trait (trait.txt) file.

## **Step 1**

The first step is to create a csv file with the input data. Here we use the \<trait\>.txt files as input to the **createCSV.py** script. The script outputs a file named **assoc.csv**.

Please run this code in the LZ_wrapper directory where the code resides.

`python createCSV.py -h` to look at the inputs needed. 

`python createCSV.py Example/t1.txt Example/t2.txt ...` 

For example: `python createCSV.py ai.txt cpd.txt si.txt sc.txt`

## **Step2**

The second step is to create the html file and the web server application needed to run LocusZoom js. The script **createCode.py** creates these files. Inputs to the script is the phenotypes.txt file. The script outputs text in html format which can be redirected to a html file of user's choice, and app.py files for the web server.

`python createCode.py -h` to look at the inputs needed. 

`python createCode.py Example/phenotypes.txt`

As a reminder, the phenotype ids (column 1) of the phenotypes file should exactly match the input file names for each trait in step 1. This step creates a html file for LocusZoom with LD for each trait/phenotype (trait.html), a multiple phenotpye LocusZoom plot for all traits together (index.html) and a **wrapper.html** which lets a user traverse between the different plots generated.
The **wrapper.html** file is stored in the templates/ directory. The **index.html** and **trait.html** files are in the static/html directory.
The **app.py** and **app_login.py** are generated in the top-level directory.

## **Step3**

The third and the final step is to run the web application generated by Step 2. There are two versions here - (1) **app.py**, which is open to anyone with the url. (2) **app_login.py**, which requires a username and password to login and access. Please choose the version most appropriate for your application. 

**Login information:** 

For app_login.py, the default username is *admin* and the default password is *password*. If you are using the generated **app_login.py**,  please go into the script (file) and modify line 22, to replace *admin* and *password* with username and password of your choice:

`if request.form['password'] == 'your password choice' and request.form['username'] == 'your username choice':`

**How to run the web server**

`python app.py`

or

`python app_login.py`

This starts a localhost server at port 8000. 


If your local port 8000 is already taken, specify another part by changing the following line in app.py before running it:

`app.run(port=8000)`

You can specify any port number there.

Be sure to also modify the html file generated in step 2 by changing the port number in the following line:

`var dataBase = 'http://127.0.0.1:8000/single/'`

by replacing the 8000 with the actual port number. Alternatively, you can modify that line in mid.html and then run createCode.py again.

## **Step 4**

Now that the app is running, access it in the web browser of your choice with:
`http://127.0.0.1:8000`

**NOTE:** The initial panel for LocusZoom for all phenotypes combined is around the `15:78800000-78820000` region of chromosome 15. If you wish to modify that, go into **index.html** file, and modify the following line:

`<div id="plot" data-region="15:78800000-78820000"></div>`

to your region of interest. For example, if a user wants the initial view to have the range 1011001-1020000 of Chromosome 1, they will replace the `15:78800000-78820000` with `1:1011001-1020000`.
